File: back-end/src/main/java/pl/straburzynski/campfiresongs/service/SongService.java
package pl.straburzynski.campfiresongs.service;

import org.springframework.stereotype.Service;
import pl.straburzynski.campfiresongs.exception.SongNotFoundException;
import pl.straburzynski.campfiresongs.model.Song;
import pl.straburzynski.campfiresongs.repository.SongRepository;

import java.util.List;
import java.util.UUID;

@Service
public class SongService {

    private final SongRepository songRepository;

    public SongService(SongRepository songRepository) {
        this.songRepository = songRepository;
    }

    public List<Song> findAll() {
        return songRepository.findAll();
    }

    public Song findById(UUID id) {
        return songRepository.findById(id).orElseThrow(() -> new SongNotFoundException(id));
    }

}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/song/service/SongService.kt
package pl.straburzynski.campfiresongs.song.service

import org.springframework.stereotype.Service
import pl.straburzynski.campfiresongs.song.exception.CannotUpdateSongException
import pl.straburzynski.campfiresongs.song.exception.SongExistsException
import pl.straburzynski.campfiresongs.song.exception.SongNotFoundException
import pl.straburzynski.campfiresongs.song.model.Song
import pl.straburzynski.campfiresongs.song.model.SongDto
import pl.straburzynski.campfiresongs.song.model.SongHeadersDto
import pl.straburzynski.campfiresongs.song.repository.SongRepository
import java.util.UUID
import kotlin.jvm.optionals.getOrNull


@Service
class SongService(
    private val songRepository: SongRepository,
    private val songConverter: SongConverter
) {
    fun create(songDto: SongDto): Song {
        val exists = songRepository.existsByAuthorLikeAndTitleLikeAllIgnoreCase(songDto.author, songDto.title)
        if (exists) throw SongExistsException(songDto)
        val savedSong = songRepository.save(
            Song(author = songDto.author, title = songDto.title, lyrics = songDto.lyrics)
        )
        return savedSong
    }

    fun findAll(): List<SongDto> = songRepository.findAll().toList()
        .map(songConverter::convertFromSong)

    fun findAllHeaders(): List<SongHeadersDto> = findAll()
        .map { songDto: SongDto ->
            SongHeadersDto(
                songDto.id,
                songDto.author,
                songDto.title
            )
        }

    fun findById(id: UUID): SongDto {
        val song = songRepository.findById(id).getOrNull() ?: throw SongNotFoundException(id.toString())
        return songConverter.convertFromSong(song)
    }

    fun update(id: UUID, songDto: SongDto): Song {
        val song = songConverter.convertFromSongDto(songDto)
        if (songRepository.existsById(id)) return songRepository.save(song)
        else throw CannotUpdateSongException(id, songDto)
    }

    fun deleteById(id: UUID) = songRepository.deleteById(id)
}